cirq/ops/eigen_gate.py

```diff
@@ -322,6 +322,32 @@ def _resolve_parameters_(self: TSelf, param_resolver) -> TSelf:
         return self._with_exponent(
                 exponent=param_resolver.value_of(self._exponent))
 
+    def _equal_up_to_global_phase_(self, other, atol):
+        if not isinstance(other, EigenGate):
```

Comment:
Should be `type(self)` instead of `EigenGate`. Otherwise you will end up comparing XPowGate to ZPowGate.

Make sure to test that XPowGate is not approximately equal to ZPowGate up to global phase.

Reply:
The method only returns `True` if the gates without exponents or phases are approximately equal, so the type check can be done there if desired. Certainly there are gates of different types that are equal up to global phase, e.g., `X ** 0` and `Z ** 0`. False negatives are fine (and I think quite common at the moment), but the change you suggest would preclude them from being addressed eventually.

(Note that I added a test for `XPowGate` and `ZPowGate` that passes without any changes to the code.)

Reply:
I don't think X**0 is approximately equal to Z**0 up to global phase. Their unitaries are equal, but the classes are distinct. If you wanted to compare the unitaries, you'd compare the unitaries.

Reply:
You're right that the existing code handles this. I think it's because of the delegating to approx_eq

Reply:
Is what "equality" of gates means in Cirq documented anywhere?